csharp/ql/src/semmle/code/csharp/dataflow/internal/FrameworkDataFlowImpl.qll

```diff
@@ -0,0 +1,184 @@
+
+  string toString() { result = r }
+}
+
+module FlowSource {
+  abstract class Range extends string {
+    bindingset[this]
+    Range() { any() }
+
+    predicate isSubsetOf(FlowSource fs) { none() }
+
+    FlowSource toFlowSource() { result = MkFlowSource(this) }
+  }
+}
+
+private newtype TFlowSink = MkFlowSink(FlowSink::Range r)
+
+/** A flow-sink specification. */
+class FlowSink extends MkFlowSink {
+  private FlowSink::Range r;
+
+  FlowSink() { this = MkFlowSink(r) }
+
+  predicate isSubsetOf(FlowSink fs) { r.isSubsetOf(fs) }
+
+  string toString() { result = r }
+}
+
+module FlowSink {
+  abstract class Range extends string {
+    bindingset[this]
+    Range() { any() }
+
+    predicate isSubsetOf(FlowSink fs) { none() }
+
+    FlowSink toFlowSink() { result = MkFlowSink(this) }
+  }
+}
+
+/** A data-flow model for a given framework. */
+abstract class FrameworkDataFlow extends string {
+  bindingset[this]
+  FrameworkDataFlow() { any() }
+
+  /**
+   * Holds if data may flow from `input` to `output` when calling `c`.
+   *
+   * `preservesValue` indicates whether this is a value-preserving step
+   * or a taint-step.
+   */
+  pragma[nomagic]
+  predicate hasSummary(
+    FrameworkCallable c, SummaryInput input, SummaryOutput output, boolean preservesValue
+  ) {
+    none()
+  }
+
+  /**
+   * Holds if data may flow from `input` to `output` when calling `c`.
+   *
+   * `inputContents` describes the contents that is popped from the access
+   * path from the input and `outputContents` describes the contents that
+   * is pushed onto the resulting access path.
+   *
+   * `preservesValue` indicates whether this is a value-preserving step
+   * or a taint-step.
+   */
+  pragma[nomagic]
+  predicate hasSummary(
+    FrameworkCallable c, SummaryInput input, ContentList inputContents, SummaryOutput output,
+    ContentList outputContents, boolean preservesValue
+  ) {
+    none()
+  }
+
+  /**
+   * Holds if the content list obtained by consing `head` onto `tail` is needed
+   * for a summary specified by `hasSummary()`.
+   *
+   * This predicate is needed for QL technical reasons only (the IPA type used
+   * to represent content lists needs to be bounded).
+   *
+   * Only summaries using content lists of length >= 2 need to override this
+   * predicate.
+   */
+  pragma[nomagic]
+  predicate requiresContentList(Content head, ContentList tail) { none() }
+
+  /**
+   * Holds if values stored inside `content` are cleared on objects passed as
+   * arguments of type `input` to calls that target `c`.
+   */
+  pragma[nomagic]
+  predicate clearsContent(FrameworkCallable c, SummaryInput input, Content content) { none() }
+
+  /** Holds if `n` is a flow source of type `source. */
+  pragma[nomagic]
+  predicate hasSource(Node n, FlowSource source) { none() }
+
+  /** Holds if `n` is a flow sink of type `sink. */
+  pragma[nomagic]
+  predicate hasSink(Node n, FlowSink sink) { none() }
```

Comment:
Those are some very valid points, perhaps something like this would work instead?
```ql
/** Holds if output of type `output` from `c` is a flow source of type `source`.  */
predicate hasSource(FrameworkCallable c, SummaryOutput output, FlowSource source) { none() }
```
and similar for `hasSink`. That, however, restricts sources and sinks to output from/input to callables, but perhaps that will cover most cases.

Reply:
LGTM